File: robotcontrol2/OccupancyGrid.swift
//
//  Occupancy.swift
//  robotcontrol2
//
//  Created by Sebastian Theophil on 20.02.15.
//  Copyright (c) 2015 Sebastian Theophil. All rights reserved.
//

import Foundation
import Cocoa
import Accelerate

protocol ScalableOffsetable {
    func +(left: Self, right: CGVector) -> Self
    func -(left: Self, right: CGVector) -> Self
    
    func *(left: Self, right: CGFloat) -> Self
    func /(left: Self, right: CGFloat) -> Self
}

// Geometry extensions
func angularDistance( angleA : CGFloat, angleB : CGFloat ) -> CGFloat {
    var angle = angleA - angleB
    // normalize angle to be between (-pi, +pi]
    while(angle < CGFloat(-M_PI)) {
        angle+=2*CGFloat(M_PI)
    }
    while(CGFloat(M_PI) <= angle) {
        angle-=2*CGFloat(M_PI)
    }
    return angle
}

// CGPoint operators
func + (left: CGPoint, right: CGVector) -> CGPoint {
    return CGPointMake(left.x + right.dx, left.y + right.dy)
}

func -(left: CGPoint, right: CGVector) -> CGPoint {
    return CGPointMake(left.x - right.dx, left.y - right.dy)
}

func *(left: CGPoint, right: CGFloat) -> CGPoint {
    return CGPointMake(left.x * right, left.y * right)
}

func /(left: CGPoint, right: CGFloat) -> CGPoint {
    return CGPointMake(left.x / right, left.y / right)
}

func -(left: CGPoint, right: CGPoint) -> CGVector {
    return CGVector(dx: left.x - right.x, dy: left.y - right.y)
}


// CGRect operators
func +(left: CGRect, right: CGVector) -> CGRect {
    return left.rectByOffsetting(dx: right.dx, dy: right.dy)
}

func -(left: CGRect, right: CGVector) -> CGRect {
    return left.rectByOffsetting(dx: -right.dx, dy: -right.dy)
}

func *(left: CGRect, right: CGFloat) -> CGRect {
    return CGRect(origin: left.origin*right, size: left.size*right)
}

func /(left: CGRect, right: CGFloat) -> CGRect {
    return CGRect(origin: left.origin/right, size: left.size/right)
}

func |=(left: CGRect, right: CGPoint) -> CGRect {
    let origin = CGPoint(
        x: min(left.minX, right.x),
        y: min(left.minY, right.y))
    
    let size = CGSize(
        width: max(left.width, right.x - origin.x),
        height: max(left.height, right.y - origin.y)
    )
    return CGRect(origin: origin, size: size)
}

// CGSize operators
func *(left: CGSize, right: CGFloat) -> CGSize {
    return CGSize(width: left.width * right, height: left.height * right)
}

func /(left: CGSize, right: CGFloat) -> CGSize {
    return CGSize(width: left.width / right, height: left.height / right)
}

// CGVector operators
func *(left: CGVector, right: CGFloat) -> CGVector {
    return CGVector(dx: left.dx * right, dy: left.dy * right)
}

func /(left: CGVector, right: CGFloat) -> CGVector {
    return CGVector(dx: left.dx / right, dy: left.dy / right)
}

extension CGVector {
    init(_ pt: CGPoint) {
        dx = pt.x
        dy = pt.y
    }
    
    init(_ sz: CGSize) {
        dx = sz.width
        dy = sz.height
    }
}

extension CGSize {
    init(_ v: CGVector) {
        width = v.dx
        height = v.dy
    }
}
    
extension CGPoint : ScalableOffsetable {
    init(fromAngle angle: CGFloat, distance: CGFloat) {
        x = distance * cos(angle)
        y = distance * sin(angle)
    }

    func distance(pt : CGPoint) -> CGFloat {
        return hypot(x - pt.x, y - pt.y)
    }
    
    func SqrAbs() -> CGFloat {
        return x*x + y*y
    }
    
    func quadrant() -> Int {
        return x < 0
            ? (y < 0 ? 2 : 1)
            : (y < 0 ? 3 : 0)
    }
    
    func rotated(fAngle: CGFloat) -> CGPoint {
        let c = cos(fAngle)
        let s = sin(fAngle)
        return CGPoint(x: x * c - y * s, y: x * s + y * c)
    }
    
    // < 0 -> v is left of this
    // > 0 -> v is right of this
    func compare(v: CGPoint) -> Int {
        let d = y * v.x - x * v.y
        return sign(d)
    }
}

func RectFromPoints(points: [CGPoint]) -> (CGPoint, CGSize) {
    let bottomLeft = CGPoint(
        x: minElement(points.map{ $0.x }),
        y: minElement(points.map{ $0.y })
    )
    let topRight = CGPoint(
        x: maxElement(points.map{ $0.x }),
        y: maxElement(points.map{ $0.y })
    )
    return (bottomLeft, CGSize(topRight - bottomLeft))
}

extension CGRect : ScalableOffsetable {
    init(fromPoints points: CGPoint...) {
        (origin, size) = RectFromPoints(points)
    }
    
    init(fromPoints points: [CGPoint]) {
        (origin, size) = RectFromPoints(points)
    }
    
    init(centeredAt pt: CGPoint, size: CGSize) {
        self.origin = pt - CGVector(size/2)
        self.size = size
    }
}

func sign<T : SignedNumberType>(x : T) -> Int {
    return x < 0 ? -1 : (0 < x ? 1 : 0)
}

func sqr<T : IntegerArithmeticType>(x : T) -> T {
    return x * x
}

struct SArc {
    var ptCenter: CGPoint
    var angleFrom: CGFloat
    var angleTo: CGFloat
    var radius: CGFloat
    
    init(_ pt: CGPoint, fromAngle: CGFloat, toAngle: CGFloat, withRadius: CGFloat) {
        self.ptCenter = pt
        self.angleFrom = fromAngle
        self.angleTo = toAngle
        self.radius = withRadius
    }
    
    func forEachPoint(foreach : (Int, Int, CGFloat) -> Void) {
        var ptFrom =  CGPoint(fromAngle: angleFrom, distance: radius) // relative to ptCenter
        var ptTo =  CGPoint(fromAngle: angleTo, distance: radius)
        
        assert(angularDistance(angleFrom, angleTo)<CGFloat(M_PI/2))
        assert(ptFrom.compare(ptTo) != 0)
        
        if ptFrom.compare(ptTo) < 0 { // ptFrom should be left of ptTo
            swap(&ptFrom, &ptTo)
        }
        
        let nQuadrantFrom = ptFrom.quadrant()
        let nQuadrantTo = ptTo.quadrant()
        
        assert(nQuadrantFrom==nQuadrantTo || nQuadrantFrom==(nQuadrantTo+1)%4)
        
        var rectBound = CGRect(fromPoints: ptFrom, ptTo, CGPoint(x: 0, y: 0))
        if nQuadrantFrom != nQuadrantTo {
            switch nQuadrantFrom {
            case 0: rectBound |= CGPoint(x: radius, y: 0)
            case 1: rectBound |= CGPoint(x: 0, y: radius)
            case 2: rectBound |= CGPoint(x: -radius, y: 0)
            case 3: rectBound |= CGPoint(x: 0, y: -radius)
            default: assert(false)
            }
        }
        
        let fSqrRadius = CGFloat(radius * radius)
        for y in Int(round(rectBound.minY)) ... Int(round(rectBound.maxY)) {
            var bFoundPointsInLine = false
            for x in Int(round(rectBound.minX)) ... Int(round(rectBound.maxX)) {
                let pt = CGPoint(x: x, y: y)
                let fSqrDistance = pt.SqrAbs()
                
                if fSqrDistance < fSqrRadius
                && 0<=ptFrom.compare(pt) // this is very strict and does not
                && 0<=pt.compare(ptTo) { // count grid cells partially inside arc
                    bFoundPointsInLine = true
               
                    foreach(x + Int(round(ptCenter.x)), y + Int(round(ptCenter.y)), fSqrDistance)
                    
                } else if bFoundPointsInLine {
                    break // go to next line
                }
            }
        }
    }
}

struct SRotatedRect {
    var ptCenter : CGPoint
    var size : CGSize
    var fAngle : CGFloat
    
    func forEachPoint(foreach : (Int, Int) -> Void) {
        let apt = [
            CGPoint(x: -size.width/2, y: -size.height/2).rotated(fAngle) + CGVector(ptCenter),
            CGPoint(x: size.width/2, y: -size.height/2).rotated(fAngle) + CGVector(ptCenter),
            CGPoint(x: size.width/2, y: size.height/2).rotated(fAngle) + CGVector(ptCenter),
            CGPoint(x: -size.width/2, y: size.height/2).rotated(fAngle) + CGVector(ptCenter)
        ]
        
        func rasterize(var ptA: CGPoint, var ptB: CGPoint, f : (Int, Int) -> Void) {
            if(ptB.x<ptA.x) {
                swap(&ptB, &ptA)
            }
            
            let nBegin = Int(round(ptA.x))
            let nEnd = Int(round(ptB.x))
            
            if(nBegin == nEnd) {
                // straight vertical line
                let nBeginY = Int( round(min(ptA.y, ptB.y)) )
                let nEndY = Int( round(max(ptA.y, ptB.y)) )
                for y in nBeginY ... nEndY {
                    f(nBegin, y)
                }
            } else {
                let m : CGFloat = (ptB.y - ptA.y) / (ptB.x - ptA.x)
                if(abs(m)<=1) { // x-step
                    for x in nBegin ... nEnd {
                        f(x, Int(round(ptA.y + m * CGFloat(x - nBegin))))
                    }
                } else { // y-step
                    if(ptB.y < ptA.y) {
                        swap(&ptB, &ptA);
                    }
                    let nBeginY = Int( round(ptA.y) )
                    let nEndY = Int( round(ptB.y) )
                    
                    for y in nBeginY ... nEndY {
                        f(Int(round(ptA.x + CGFloat(y - nBeginY) / m)), y)
                    }
                }
            }
        }
        
        // TODO: The map could be avoided by sorting the line segments
        var dictMinMaxX = [Int: (Int, Int)]()
        for i in 0 ..< apt.count {
            rasterize(apt[i], apt[(i+1)%apt.count], { (x: Int, y: Int) -> Void in
                if let pairMinMax = dictMinMaxX[y] {
                    dictMinMaxX[y] = (min(pairMinMax.0, x), max(pairMinMax.1, x))
                } else {
                    dictMinMaxX[y] = (x, x)
                }
            })
        }
        
        for (y, pairMinMaxX) in dictMinMaxX {
            for x in pairMinMaxX.0 ... pairMinMaxX.1 {
                foreach(x, y)
            }
        }
    }
}

class OccupancyGrid {
    init() {
        // occupancy grid in log odds representation
        grid = [[Double]](count: extent,
            repeatedValue: [Double](count: extent, repeatedValue: 0.0))
        
        // occupancy grid as 8-bit greyscale image
        // (0,0) is top-left, different from other Cocoa coordinate systems
        image = NSBitmapImageRep(bitmapDataPlanes: nil,
            pixelsWide: extent, pixelsHigh: extent,
            bitsPerSample: 8,
            samplesPerPixel: 1,
            hasAlpha: false,
            isPlanar: true,
            colorSpaceName: NSDeviceWhiteColorSpace,
            bitmapFormat: NSBitmapFormat.allZeros,
            bytesPerRow: 0, bitsPerPixel: 0)!

        imageEroded = image.copy() as! NSBitmapImageRep
        
        var color = 128
        for x in 0 ..< Int(image.size.width) {
            for y in 0 ..< Int(image.size.height) {
                image.setPixel(&color, atX: x, y: y)
            }
        }
    }
    
    func update(pt: CGPoint, fYaw: CGFloat, nAngle : Int16, nDistance : Int16 ) {
        // update occgrid & occgridImage
        assert(nAngle==0 || abs(nAngle)==90)
        let fAngleSonar = CGFloat(fYaw) + CGFloat(M_PI_2) * CGFloat(sign(nAngle))
        
        let arc = SArc(transform(pt),
            fromAngle: fAngleSonar - sonarOpeningAngle/2,
            toAngle: fAngleSonar + sonarOpeningAngle/2,
            withRadius: (CGFloat(nDistance) + sonarDistanceTolerance/2)/scale)
        
        let fSqrMaxDistance = pow(sonarMaxDistance/scale, 2)
        let fSqrMeasuredDistance = pow((CGFloat(nDistance) - sonarDistanceTolerance/2)/scale, 2)
        
        arc.forEachPoint({ (x: Int, y: Int, fSqrDistance: CGFloat) -> Void in
            if fSqrDistance < fSqrMaxDistance {
                let fInverseSensorModel = fSqrDistance < fSqrMeasuredDistance
                    ? -0.5 // free
                    : Double(100.0 / self.scale) / sqrt(Double(fSqrDistance)) // occupied
                self.setGrid(x, y, self.grid[x][y] + fInverseSensorModel) // - prior which is 0
            }
        })
        
        // Clear position of robot itself
        let rectRobot = SRotatedRect(ptCenter: transform(pt), size: sizeRobot/scale, fAngle: fYaw)
        rectRobot.forEachPoint({ (x: Int, y:Int) -> Void in
            self.setGrid(x, y, -100)
        })
        
        // Erode image
        // A pixel p in imageEroded is marked free when the robot centered at p does not occupy an occupied pixel in self.image
        // i.e. the pixel p has the maximum value of the surrounding pixels inside the diameter defined by the robot's size
        // We overestimate robot size by taking robot diagonal 
        let nKernelDiameter = UInt( ceil( sqrt( pow(sizeRobot.width, 2) + pow(sizeRobot.height, 2) ) / scale ) )
        let anKernel = [UInt8](count: Int(nKernelDiameter * nKernelDiameter), repeatedValue: 0)
        var vimgbufInput = vImage_Buffer(data: image.bitmapData, height: UInt(image.pixelsHigh), width: UInt(image.pixelsWide), rowBytes: image.bytesPerRow)
        var vimgbufOutput = vImage_Buffer(data: imageEroded.bitmapData, height: UInt(image.pixelsHigh), width: UInt(image.pixelsWide), rowBytes: image.bytesPerRow)
        
        vImageErode_Planar8( &vimgbufInput, &vimgbufOutput, 0, 0, anKernel, nKernelDiameter, nKernelDiameter, UInt32(kvImageNoFlags) )
    }
    
    func setGrid(x: Int, _ y: Int, _ value: Double) {
        grid[x][y] = value
        var color = Int(round(1 / ( 1 + exp( value ))  * 255))
        image.setPixel(&color, atX: x, y: y)
    }
    
    func draw(bShowOriginalMap : Bool) {
        let sizeImage = image.size * scale
        (bShowOriginalMap ? image : imageEroded).drawInRect(NSRect(
            x: -sizeImage.width/2,
            y: sizeImage.height/2,
            width: sizeImage.width,
            height: -sizeImage.height
            ))
    }
    
    // TODO: integer point, integer intervals?
    func closestUnknownPoint(ptCenter: CGPoint) -> CGPoint {
        let intvlUnknown = (UInt8(0.4 * 255), UInt8(0.6 * 255))
        
        let ptCenterTransformed = transform(ptCenter)
        let ptnCenter = (Int(round(ptCenterTransformed.x)), Int(round(ptCenterTransformed.y)))
        var ptnMin = ptnCenter
        var nSqrMinDistance = Int.max
        
        // nDistanceMax is inclusive
        var nDistanceMax = max(extent - ptnCenter.0 - 1, ptnCenter.0,
            extent - ptnCenter.1 - 1, ptnCenter.1)
        
        let anBitmap = image.bitmapData
        let cbBytesPerRow = image.bytesPerRow
        
        let UpdateClosest = { (x: Int, y: Int, nDistance: Int) -> Void in
            let nColor = anBitmap[y * cbBytesPerRow + x]
            if intvlUnknown.0 <= nColor && nColor < intvlUnknown.1 {
                
                let nSqrDistance = sqr(x - ptnCenter.0) + sqr(y - ptnCenter.1)
                if nSqrDistance < nSqrMinDistance {
                    // Point x, y is on the boundary of a rectangle around ptCenter
                    // When we find a new closest point, calculate the size of the largest rectangle
                    // that contains all points that may be closer than x,y
                    nDistanceMax = Int(ceil(sqrt(CGFloat(nSqrDistance)) / CGFloat(nDistance)))
                    
                    ptnMin = (x, y)
                    nSqrMinDistance = nSqrDistance
                }
            }
        }
        
        for nDistance in 1 ... nDistanceMax {
            let intvlnX = (max(0, ptnCenter.0-nDistance), min(extent-1, ptnCenter.0+nDistance))
            for x in intvlnX.0 ... intvlnX.1 {
                if 0 <= ptnCenter.1 - nDistance {
                    UpdateClosest(x, ptnCenter.1 - nDistance, nDistance)
                }
                if ptnCenter.1 + nDistance < extent {
                    UpdateClosest(x, ptnCenter.1 + nDistance, nDistance)
                }
            }
            
            let intvlnY = (max(0, ptnCenter.1-nDistance), min(extent-1, ptnCenter.1+nDistance))
            for y in intvlnY.0 ... intvlnY.1 {
                if 0 <= ptnCenter.0 - nDistance {
                    UpdateClosest(ptnCenter.0 - nDistance, y, nDistance)
                }
                if ptnCenter.0 + nDistance < extent {
                    UpdateClosest(ptnCenter.0 + nDistance, y, nDistance)
                }
            }
        }
        
        return inverseTransform(CGPoint(x: ptnMin.0, y: ptnMin.1))
    }
    
    private func transform<T: ScalableOffsetable>(geom: T) -> T { // to pixel coordinates
        let gridOffset = CGVector(dx: extent/2, dy: extent/2)
        return (geom / scale) + gridOffset
    }
    
    private func inverseTransform<T: ScalableOffsetable>(geom: T) -> T { // from pixel coordinate
        let gridOffset = CGVector(dx: extent/2, dy: extent/2)
        return (geom - gridOffset) * scale
    }
    
    // When transforming coordinates to grid coordinates, always round to nearest
    // i.e., the grid cell (0,0) is centered at point (0,0)
    let extent = 1000 // grid size in pixel, even number
    let scale : CGFloat = 5 // cm per pixel
    
    // occupancy grid in log odds representation
    var grid : [[Double]]
    
    // occupancy grid as 8-bit greyscale image
    // (0,0) is top-left, different from other Cocoa coordinate systems
    var image : NSBitmapImageRep
    var imageEroded : NSBitmapImageRep
}

File: robotcontrol2/occupancy_grid.cpp
//
//  occupancy_grid.cpp
//  robotcontrol2
//
//  Created by Sebastian Theophil on 24.09.15.
//  Copyright © 2015 Sebastian Theophil. All rights reserved.
//

#include "occupancy_grid.h"
#import "../rover/src/rover.h" // Data structures and configuration data shared with Arduino controller

#include <assert.h>
#include <cstdlib>
#include <cmath>
#include <algorithm>
#include <unordered_map>

#include <boost/range/size.hpp>
#include <opencv2/imgproc.hpp>

namespace rbt {
    double angularDistance( double fAngleA, double fAngleB) {
        auto fAngle = fAngleA - fAngleB;
        
        // normalize angle to be between (-pi, +pi]
        while(fAngle < -M_PI) {
            fAngle+=2*M_PI;
        }
        while(M_PI <= fAngle) {
            fAngle-=2*M_PI;
        }
        assert(-M_PI<=fAngle && fAngle<M_PI);
        return fAngle;
    }
        
    struct SArc {
        rbt::point<int> m_ptnCenter;
        double m_fAngleFrom;
        double m_fAngleTo;
        double m_fRadius;
        
        template<class Func>
        void for_each_pixel(Func foreach) const {
            // TODO: Approximate arc with lines, use cv::LineIterator instead
            auto szFrom = rbt::size<int>::fromAngleAndDistance(m_fAngleFrom, m_fRadius);
            auto szTo = rbt::size<int>::fromAngleAndDistance(m_fAngleTo, m_fRadius);
            
            assert(angularDistance(m_fAngleFrom, m_fAngleTo)<M_PI/2);
            // This can happen because szFrom and szTo are already cast to int coordinates:
            // assert(szFrom.compare(szTo) != 0);
            
            if(szFrom.compare(szTo) < 0) { // ptFrom should be left of ptTo
                std::swap(szFrom, szTo);
            }
            
            auto nQuadrantFrom = szFrom.quadrant();
            auto nQuadrantTo = szTo.quadrant();
            
            assert(nQuadrantFrom==nQuadrantTo || nQuadrantFrom==(nQuadrantTo+1)%4);
            
            auto rectnBound = rbt::rect<int>::bound({rbt::point<int>::zero(),
                                                    rbt::point<int>::zero() + szFrom,
                                                    rbt::point<int>::zero() + szTo});
            if(nQuadrantFrom != nQuadrantTo) {
                auto nRadius = rbt::numeric_cast<int>(m_fRadius);
                switch(nQuadrantFrom) {
                    case 0: rectnBound |= rbt::point<int>(nRadius, 0); break;
                    case 1: rectnBound |= rbt::point<int>(0, nRadius); break;
                    case 2: rectnBound |= rbt::point<int>(-nRadius, 0); break;
                    case 3: rectnBound |= rbt::point<int>(0, -nRadius); break;
                    default: assert(false);
                }
            }
            
            auto fSqrRadius = rbt::sqr(m_fRadius);
            for(int y = rectnBound.bottom; y <= rectnBound.top; ++y) { // both-inclusive
                // Scan bound rect lines in y-direction for interval between vectors szFrom and szTo
                // Since arcs are convex, there is (exactly) one contiguous sequence of pixels
                // that fall into the arc. Therefore, we can stop processing a line once we
                // find the first pixel _not_ in the arc, after we have found pixels in the arc.
                bool bFoundPointsInLine = false;
                for(int x = rectnBound.left; x <= rectnBound.right; ++x) {
                    rbt::size<int> sz(x, y); // still relative to point(0,0), the arc center
                    auto nSqrDistance = sz.SqrAbs();
                    
                    if(nSqrDistance < fSqrRadius
                    && 0<=szFrom.compare(sz)    // this is very strict and does not
                    && 0<=sz.compare(szTo)) {   // count grid cells partially inside arc
                        bFoundPointsInLine = true;
                        
                        foreach(m_ptnCenter + sz, nSqrDistance);
                    } else if(bFoundPointsInLine) {
                        break; // go to next line
                    }
                }
            }
        }
    };
    
    struct SRotatedRect {
        rbt::point<int> m_ptnCenter;
        rbt::size<double> m_szf;
        double m_fAngle;
        
        template<typename Func>
        void for_each_pixel(Func foreach) {
            // Rotate the scaled rectangle.
            // TODO: For numerical precision, it may be better to rotate the rect
            // in world coordinates and then scale.
            // TODO: Use cv::LineIterator instead
            rbt::point<int> apt[] = {
                m_ptnCenter - rbt::size<int>((m_szf/2).rotated(m_fAngle)),
                m_ptnCenter + rbt::size<int>((rbt::size<double>(m_szf.x, -m_szf.y)/2).rotated(m_fAngle)),
                m_ptnCenter + rbt::size<int>((m_szf/2).rotated(m_fAngle)),
                m_ptnCenter + rbt::size<int>((rbt::size<double>(-m_szf.x, m_szf.y)/2).rotated(m_fAngle))
            };
            
            auto rasterize = [](rbt::point<int> ptA, rbt::point<int> ptB, auto foreach) {
                if(ptA.x == ptB.x) {
                    // straight vertical line
                    auto nMin = std::min(ptA.y, ptB.y);
                    auto nMax = std::max(ptA.y, ptB.y);
                    
                    for(int y = nMin; y <= nMax; ++y) {
                        foreach(ptA.x, y);
                    }
                } else {
                    if(ptB.x<ptA.x) {
                        std::swap(ptB, ptA);
                    }
                    
                    auto m = rbt::numeric_cast<double>(ptB.y - ptA.y) / (ptB.x - ptA.x);
                    if(std::abs(m)<=1) { // x-step
                        for(int x = ptA.x; x<= ptB.x; ++x) {
                            foreach(x, rbt::numeric_cast<int>(ptA.y + m * (x - ptA.x)));
                        }
                    } else { // y-step
                        auto nMin = std::min(ptA.y, ptB.y);
                        auto nMax = std::max(ptA.y, ptB.y);
                        for(int y = nMin; y <= nMax; ++y) {
                            foreach(rbt::numeric_cast<int>(ptA.x + (y - ptA.y) / m), y);
                        }
                    }
                }
            };
            
            // TODO: The map could be avoided by sorting the line segments
            std::unordered_map<int, rbt::interval<int>> mapnintvlX;
            for(int i=0; i<boost::size(apt); ++i) {
                rasterize(apt[i], apt[(i+1)%boost::size(apt)], [&](int x, int y) {
                    auto pairitb = mapnintvlX.emplace(y, rbt::interval<int>(x, x));
                    if(!pairitb.second) {
                        pairitb.first->second |= x;
                    }
                });
            }
            boost::for_each(mapnintvlX, [&](auto const& pairnintvlX) {
                for(int x = pairnintvlX.second.begin; x <= pairnintvlX.second.end; ++x) {
                    foreach(rbt::point<int>(x, pairnintvlX.first));
                }
            });
        }
    };

    COccupancyGrid::COccupancyGrid(rbt::size<int> const& szn, int nScale)
    :   m_szn(szn), m_nScale(nScale),
        m_matfMapLogOdds(m_szn.x, m_szn.y, CV_32FC1, 0.0f),
        m_matnMapGreyscale(m_szn.x, m_szn.y, CV_8UC1, 128),
        m_matnMapEroded(m_szn.x, m_szn.y, CV_8UC1, 128)
    {
        assert(0==szn.x%2 && 0==szn.y%2);
    }

    
    void COccupancyGrid::update(point<double> const& ptf, double fYaw, int nAngle, int nDistance) {
        assert(nAngle==0 || std::abs(nAngle)==90);
        auto const fAngleSonar = fYaw + M_PI_2 * rbt::sign(nAngle);
        auto const ptnGrid = toGridCoordinates(ptf);
        
        auto const fSqrMaxDistance = rbt::sqr(c_fSonarMaxDistance/m_nScale);
        auto const fSqrMeasuredDistance = rbt::sqr((nDistance - c_fSonarDistanceTolerance/2)/m_nScale);
        
        auto UpdateMap = [this](rbt::point<int> const& pt, float fValue) {
            m_matfMapLogOdds.at<float>(pt.y, pt.x) = fValue;
            auto const nColor = rbt::numeric_cast<std::uint8_t>(1.0 / ( 1.0 + std::exp( fValue )) * 255);
            m_matnMapGreyscale.at<std::uint8_t>(pt.y, pt.x) = nColor;
        };
        
        SArc arc{ptnGrid,
            fAngleSonar - c_fSonarOpeningAngle/2,
            fAngleSonar + c_fSonarOpeningAngle/2,
            (nDistance + c_fSonarDistanceTolerance/2)/m_nScale
        };
        arc.for_each_pixel([&](point<int> const& pt, double fSqrDistance) {
            if(fSqrDistance < fSqrMaxDistance) {
                auto const fInverseSensorModel = fSqrDistance < fSqrMeasuredDistance
                    ? -0.5 // free
                    : (100.0 / m_nScale) / std::sqrt(fSqrDistance); // occupied

                UpdateMap(pt, m_matfMapLogOdds.at<float>(pt.y, pt.x) + fInverseSensorModel); // - prior which is 0
            }
        });
        
        // Clear position of robot itself
        SRotatedRect rectRobot{ptnGrid, rbt::size<double>(c_nRobotWidth, c_nRobotHeight)/m_nScale, fYaw};
        rectRobot.for_each_pixel([&](rbt::point<int> const& pt) { UpdateMap(pt, -100); });
        
        // Erode image
        // A pixel p in imageEroded is marked free when the robot centered at p does not occupy an occupied pixel in self.image
        // i.e. the pixel p has the maximum value of the surrounding pixels inside the diameter defined by the robot's size
        // We overestimate robot size by taking robot diagonal        
        static const int s_nKernelDiameter =
            rbt::numeric_cast<int>(std::ceil(std::sqrt(rbt::size<int>(c_nRobotWidth, c_nRobotHeight).SqrAbs()) / m_nScale));
        static const cv::Mat s_matnKernel = cv::Mat(s_nKernelDiameter, s_nKernelDiameter, CV_8UC1, 1);
        cv::erode(m_matnMapGreyscale, m_matnMapEroded, s_matnKernel);
    }
    
    point<int> COccupancyGrid::toGridCoordinates(point<double> const& pt) const {
        return point<int>(pt/m_nScale) + m_szn/2;
    }

    point<int> COccupancyGrid::toWorldCoordinates(point<int> const& pt) const {
        return (pt - m_szn/2) * m_nScale;
    }
}
